magma/array.py

```diff
@@ -303,6 +303,13 @@ def _is_whole_slice(self, key):
                  key[-1] == slice(None, len(self)) or
                  key[-1] == slice(0, len(self))))
 
+    def _is_valid_slice(self, key):
+        start, stop = key.start, key.stop
+        if (((start is not None and (start >= self.N or start <= -self.N))) or
+                (stop is not None and (stop > self.N or stop < -self.N))):
+            return False
+        return True
+
```

Comment:
Prefer free function rather than member function, and pushed negate into condition, and direct return rather than `if`. Can be invoked as `_is_valid_slide(T.N, key)`.
```suggestion
def _is_valid_slice(n, key):
    start, stop = key.start, key.stop
    return ((start is None or start in range(-n + 1, n)) and
            (stop is None or stop not in range(-n, n + 1)))

```

Reply:
Using `in range` seems like a bad choice for performance, I think python will expand to the range to check the condition?

Here's a quick experiment for a small n
```
>>> timeit.timeit('3 in range(-64, 65)')
0.21609285899999975
>>> timeit.timeit('3 < -64 or 3 >= 65')
0.04527043200000236
```

Reply:
Interesting, would've presumed range doesn't eagerly expand for bounds checks. Thanks for verifying.

Fine using explicit comparisons over `range` but still prefer the remaining suggested changes.